FEATURE(SDS-Docker): [OS-524] Allow to Provide IAM Policies Rules

dduportal opened the pull request:
FEATURE(SDS-Docker): [OS-524] Allow to Provide IAM Policies Rules

 ##### SUMMARY

- Setting the environment variable `$IAM_RULES_FILE` to a custom IAM json file enable IAM in the pipeline using these custom rules (JSON file must be mounted in the container of course)

 ##### IMPACT

- N.C.

 ##### ADDITIONAL INFORMATION

Depends on #119 and #120 

Signed-off-by: dduportal <damien.duportal@openio.io>


[OS-524]: https://openio.atlassian.net/browse/OS-524

Commit:
FEATURE(Docker-SDS): ✨ Allow to provide IAM custom rules 🔏

 ##### SUMMARY

- Add IAM handling in entrypoint, detected through a file (path specified by an env. var)

 ##### IMPACT

- N/A

 ##### ADDITIONAL INFORMATION

Signed-off-by: dduportal <damien.duportal@openio.io>

cdelgehier commented on openio-sds/18.10/openio-docker-init.sh:
```diff
@@ -150,6 +150,23 @@ function keystone_config(){
+  if [ -f "${IAM_RULES_FILE}" ]; then
+    echo "IAM rule file found at ${IAM_RULES_FILE}: setting up IAM authorization."
+
+    # Insert "iam" before swift3 in pipeline's setup
```
I think we can activate the IAM as default and add these definitions in the docker image inventory
```
openio_oioswift_users:
  - name: "demo:demo"
    password: "DEMO_PASS"
    roles:
      - admin

openio_oioswift_iam_policy:
  'demo:demo':
    Statement:
      - Sid: FullAccess
        Action:
          - 's3:*'
        Effect: Allow
        Resource:
          - '*'
```

You will have to play with the json file to modify the ACLs definition.

cdelgehier replied:
And change the pipeline to
```
 openio_oioswift_pipeline: "{{ pipeline_tempauth_iam }}"
```

dduportal replied:
I have some concerns about expressing a default set of IAM policies:

- It feels weird to have a data structure expressed in YAML by default, that is overidden by a JSON struct (even though YAML is a super set of JSON). 

- How does it work when you provide custom rules: is it expected to be merged ? Or to override the default policy set?

But the idea of providing a default set with the system enabled is really appealing!

cdelgehier replied:
* the yaml structure is only there for initial deployment (the general case)
* the user could provide his personal acl policy via environnement vars
```
environment :
   FOOBAR : '{"foo": "bar"}'
```
```
{
  "demo:demo": {
    "Statement": [
      {
        "Action": [
          "s3:*"
        ],
        "Effect": "Allow",
        "Resource": [
          "*"
        ],
        "Sid": "FullAccess"
      }
    ]
  },
  "mytenant:user1": {
    "Statement": [
      {
        "Action": [
          "s3:*"
        ],
        "Effect": "Allow",
        "Resource": [
          "*"
        ],
        "Sid": "FullAccess"
      }
    ]
  }
}
```

dduportal replied:
mmmh, I got the following thoughts

- As you saw, we intend to provide the IAM rules through a JSON file in the container, because providing JSON data structure through shell environment variable is a rabbit hole with tons of escapements to be made, and as a general rule of thumb should be avoided at all cost unless you are a shell master who has time to waste on this 😅

- Enabling IAM by default is a breaking change. I don't know how our Dockr images are used, and I agree that it is not a "production case". But if we do so, I would suggest to enable it by default only on 20.04, and keep 19.x (and 18.x eventually) with the current "IAM opt-in" behavior.

- Agreed with the default IAM rules, I understand it better with the explanation. I still have a case I'm not sure: If I launch a container (with IAM enabled by default) with this command:
`docker run -e SWIFT_CREDENTIALS="uberadmin:uberadmin" openio/sds:20.04`,
I understand that the default rule either:
  - Apply to the user `demo`, so my user `uberadmin` is not authroized to perform any action?
  - Apply to everyone: is it possible to set the target user to `*:*` ? And if yes, will this ACL be overriden by providing the IAM rule file ( e.g. `docker run -e SWIFT_CREDENTIALS="uberadmin:uberadmin" -e IAM_RULES_FILE='/etc/iam_rules.json' -v "$(pwd)/iam_rules.json":/etc/iam_rules.json:ro openio/sds:20.04`)? Or will it be merged ?

cdelgehier reviewed (changes requested):
IAM is available in releases:
- 19.04
- 20.04
- 19.10 ? 
@murlock

murlock commented on openio-sds/19.04/openio-docker-init.sh:
```diff
@@ -151,6 +151,23 @@ function keystone_config(){
   fi
 }
 
+function iam_config(){
```
we forgot to add `s3_acl_openbar` (not sure if it is really an issue since we have only one user here)

cdelgehier replied:
It s possible to define many users
`-e SWIFT_CREDENTIALS="admin:admin:admin:.admin .reseller_admin,test2:tester2:testing2:.admin,test5:tester5:testing5:service,test:tester:testing:.admin,test:tester3:testing3"`

dduportal replied:
@cdelgehier if the user provides its own list of users, and enable IAM, then should we assume that an IAM rule must be provided?
Or should we test if no IAM rule file, then parse all users and set openbar acl to them ?

cdelgehier commented on openio-sds/README.md:
````diff
@@ -72,51 +71,71 @@ To access the `openio` command line, you can access inside the container (suppos
+- And provide the environment variable `$IAM_RULES_FILE` to specify the path to the `iam_rules.json` in the container.
+
+```console
+# docker run -d -e IAM_RULES_FILE='/etc/iam_rules.json' -v "$(pwd)/etc/iam_rules.json":/etc/iam_rules.json:ro openio/sds
````
YES !

cdelgehier replied:
In the example, you should differentiate the paths. It's easier to understand than `$(pwd)/iam_rules.json` is that on the client side (and `/etc/iam_rules.json` in the container)

cdelgehier commented on openio-sds/README.md:
````diff
@@ -72,51 +71,71 @@ To access the `openio` command line, you can access inside the container (suppos
 # docker run -d --net=host -e KEYSTONE_ENABLED='true' -e KEYSTONE_URI='192.168.56.102:5000' -e KEYSTONE_URL='192.168.56.102:35357' openio/sds
 ```
 
+-e KEYSTONE_ENABLED='true' -e KEYSTONE_URI='172.17.0.3:5000' -e KEYSTONE_URL='172.17.0.3:35357'
````
🤔

Commit:
BUGFIX/MEDIUM(OpenIO): 🐛 📝 Fix READMEs for 18.10, 19.04 and 18.10

 ##### SUMMARY

- Since https://github.com/open-io/dockerfiles/pull/121, the README for the 20.04 is fixed.
  This change transplant the fixes to the READMEs of the version 18.10, 19.04 and 19.10.

 ##### IMPACT

- N/A.

 ##### ADDITIONAL INFORMATION

Signed-off-by: dduportal <damien.duportal@openio.io>

Commit:
TEST(SDS): 🔎 🐳 🔀 switch Docker Image test to multi suites with docker-compose

 ##### SUMMARY

- Test suites are now defined by a directory each, which contains a `docker-compose.yml` file
- Multiple test suites are run sequentially

 ##### IMPACT

- N/A

 ##### ADDITIONAL INFORMATION

Signed-off-by: dduportal <damien.duportal@openio.io>

cdelgehier commented:
status ?

dduportal commented:
Aucune idée. Ping @murlock  est-ce que ce besoin existe toujours du côté dev ?

murlock commented:
If user provide list of users **and** IAM, please enable `s3_acl_openbar`

cdelgehier commented:
> If user provide list of users **and** IAM, please enable `s3_acl_openbar`

@dduportal this is an attribute of the middleware `swift3`

https://github.com/open-io/ansible-role-openio-oioswift/blob/20.04/defaults/main.yml#L175
